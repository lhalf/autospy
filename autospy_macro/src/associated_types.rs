use crate::attribute;
use proc_macro2::TokenStream;
use syn::{Ident, ItemTrait, TraitItem, TraitItemType};

#[derive(Clone)]
pub struct AssociatedType {
    pub name: Ident,
    pub r#type: TokenStream,
}

pub fn get_associated_types(item_trait: &ItemTrait) -> Option<AssociatedType> {
    item_trait
        .items
        .iter()
        .find_map(associated_types)
        .and_then(associated_type_from_attribute)
}

fn associated_types(item: &TraitItem) -> Option<&TraitItemType> {
    match item {
        TraitItem::Type(trait_type) => Some(trait_type),
        _ => None,
    }
}

fn associated_type_from_attribute(trait_item: &TraitItemType) -> Option<AssociatedType> {
    let r#type = attribute::associated_type(&trait_item.attrs)?.clone();
    Some(AssociatedType {
        name: trait_item.ident.clone(),
        r#type,
    })
}
