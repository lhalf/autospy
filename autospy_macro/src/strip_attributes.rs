use crate::{edit, inspect};
use syn::{Attribute, ItemTrait, Signature, TraitItem};

pub fn strip_attributes(mut item_trait: ItemTrait) -> ItemTrait {
    item_trait
        .items
        .iter_mut()
        .for_each(strip_attributes_from_item);
    item_trait
}

fn strip_attributes_from_item(item: &mut TraitItem) {
    match item {
        TraitItem::Fn(function) => {
            strip_autospy_attributes(&mut function.attrs);
            strip_attributes_from_signature(&mut function.sig);
        }
        TraitItem::Type(_type) => strip_autospy_attributes(&mut _type.attrs),
        _ => todo!(),
    }
}

fn strip_autospy_attributes(attributes: &mut Vec<Attribute>) {
    attributes.retain(|attribute| !inspect::is_autospy_attribute(attribute));
}

pub fn strip_attributes_from_signature(signature: &mut Signature) {
    for argument in edit::non_self_signature_arguments_mut(signature) {
        strip_autospy_attributes(&mut argument.attrs);
    }
}
