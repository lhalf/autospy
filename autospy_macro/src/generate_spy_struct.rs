use crate::associated_types::AssociatedSpyTypes;
use crate::{attribute, edit, generate, inspect};
use proc_macro2::TokenStream;
use quote::{ToTokens, format_ident, quote};
use syn::visit_mut::VisitMut;
use syn::{ItemStruct, ItemTrait, ReturnType, TraitItemFn, Type, TypeImplTrait};

pub fn generate_spy_struct(
    item_trait: &ItemTrait,
    associated_spy_types: &AssociatedSpyTypes,
) -> ItemStruct {
    let visibility = &item_trait.vis;
    let spy_name = format_ident!("{}Spy", item_trait.ident);
    let spy_fields = generate_spy_fields(item_trait, associated_spy_types);

    syn::parse2(quote! {
        #[cfg(any(test, not(feature = "test")))]
        #[derive(Default, Clone)]
        #visibility struct #spy_name {
            #(#spy_fields),*
        }
    })
    .expect("invalid generated spy struct")
}

fn generate_spy_fields(
    item_trait: &ItemTrait,
    associated_spy_types: &AssociatedSpyTypes,
) -> impl Iterator<Item = TokenStream> {
    inspect::trait_functions(item_trait)
        .map(|function| function_as_spy_field(function, associated_spy_types))
}

fn function_as_spy_field(
    function: &TraitItemFn,
    associated_spy_types: &AssociatedSpyTypes,
) -> TokenStream {
    let function_name = &function.sig.ident;

    let function = replace_associated_types(function.clone(), associated_spy_types);

    let spy_argument_type =
        generate::tuple_or_single(inspect::spyable_arguments(&function).map(argument_spy_type));

    let return_type = function_return_type(&function);

    quote! {
        pub #function_name: autospy::SpyFunction<#spy_argument_type, #return_type>
    }
}

fn replace_associated_types(
    mut function: TraitItemFn,
    associated_spy_types: &AssociatedSpyTypes,
) -> TraitItemFn {
    edit::AssociatedTypeReplacer {
        associated_spy_types,
    }
    .visit_trait_item_fn_mut(&mut function);
    function
}

fn argument_spy_type(argument: inspect::SpyableArgument) -> TokenStream {
    if let Some(into_type) = argument.into_type {
        return quote! { #into_type };
    }

    let dereferenced_type = &argument.dereferenced_type;
    match argument.dereferenced_type {
        Type::ImplTrait(TypeImplTrait { bounds, .. }) => quote! { Box<dyn #bounds> },
        _ => quote! { <#dereferenced_type as ToOwned>::Owned },
    }
}

fn function_return_type(function: &TraitItemFn) -> TokenStream {
    // specifying the return attribute takes precedence over associated type
    if let Some(specified_return_type) = attribute::return_type(&function.attrs) {
        return specified_return_type.to_token_stream();
    }

    match &function.sig.output {
        ReturnType::Default => quote! { () },
        ReturnType::Type(_arrow, return_type) => return_type.to_token_stream(),
    }
}

#[cfg(test)]
mod tests {
    use crate::associated_types::AssociatedSpyTypes;
    use crate::generate_spy_struct::generate_spy_struct;
    use proc_macro2::TokenStream;
    use quote::quote;
    use syn::{ItemStruct, ItemTrait};

    #[test]
    fn empty_generated_struct() {
        let input: ItemTrait = syn::parse2(quote! {
            trait Example {}
        })
        .unwrap();

        let expected: ItemStruct = syn::parse2(quote! {
            #[cfg(any(test, not(feature = "test")))]
            #[derive(Default, Clone)]
            struct ExampleSpy {}
        })
        .unwrap();

        assert_eq!(
            expected,
            generate_spy_struct(&input, &AssociatedSpyTypes::new())
        );
    }

    #[test]
    fn generated_spy_struct_retains_trait_visibility() {
        let input: ItemTrait = syn::parse2(quote! {
            pub trait Example {
                fn foo(&self);
            }
        })
        .unwrap();

        let expected: ItemStruct = syn::parse2(quote! {
            #[cfg(any(test, not(feature = "test")))]
            #[derive(Default, Clone)]
            pub struct ExampleSpy {
                pub foo: autospy::SpyFunction<(), ()>
            }
        })
        .unwrap();

        assert_eq!(
            expected,
            generate_spy_struct(&input, &AssociatedSpyTypes::new())
        );
    }

    #[test]
    fn generated_spy_struct_handles_owned_return_values() {
        let input: ItemTrait = syn::parse2(quote! {
            trait Example {
                fn foo(&self) -> String;
            }
        })
        .unwrap();

        let expected: ItemStruct = syn::parse2(quote! {
            #[cfg(any(test, not(feature = "test")))]
            #[derive(Default, Clone)]
            struct ExampleSpy {
                pub foo: autospy::SpyFunction<(), String>
            }
        })
        .unwrap();

        assert_eq!(
            expected,
            generate_spy_struct(&input, &AssociatedSpyTypes::new())
        );
    }

    #[test]
    fn generated_spy_struct_captures_reference_arguments_as_owned() {
        let input: ItemTrait = syn::parse2(quote! {
            trait Example {
                fn foo(&self, argument: &str);
            }
        })
        .unwrap();

        let expected: ItemStruct = syn::parse2(quote! {
            #[cfg(any(test, not(feature = "test")))]
            #[derive(Default, Clone)]
            struct ExampleSpy {
                pub foo: autospy::SpyFunction< < str as ToOwned > :: Owned , () >
            }
        })
        .unwrap();

        assert_eq!(
            expected,
            generate_spy_struct(&input, &AssociatedSpyTypes::new())
        );
    }

    #[test]
    fn generated_spy_struct_captures_associated_type_arguments() {
        let input: ItemTrait = syn::parse2(quote! {
            trait Example {
                #[autospy(String)]
                type Item;
                fn foo(&self, argument: Self::Item);
            }
        })
        .unwrap();

        let expected: ItemStruct = syn::parse2(quote! {
            #[cfg(any(test, not(feature = "test")))]
            #[derive(Default, Clone)]
            struct ExampleSpy {
                pub foo: autospy::SpyFunction< < String as ToOwned > :: Owned , () >
            }
        })
        .unwrap();

        assert_eq!(
            expected,
            generate_spy_struct(
                &input,
                &associated_spy_types(quote! { Item }, quote! { String })
            )
        );
    }

    #[test]
    fn generated_spy_struct_handles_associated_type_returns() {
        let input: ItemTrait = syn::parse2(quote! {
            trait Example {
                #[autospy(String)]
                type Item;
                fn foo(&self) -> Self::Item;
            }
        })
        .unwrap();

        let expected: ItemStruct = syn::parse2(quote! {
            #[cfg(any(test, not(feature = "test")))]
            #[derive(Default, Clone)]
            struct ExampleSpy {
                pub foo: autospy::SpyFunction< (), String >
            }
        })
        .unwrap();

        assert_eq!(
            expected,
            generate_spy_struct(
                &input,
                &associated_spy_types(quote! { Item }, quote! { String })
            )
        );
    }

    fn associated_spy_types(ident: TokenStream, r#type: TokenStream) -> AssociatedSpyTypes {
        [(syn::parse2(ident).unwrap(), syn::parse2(r#type).unwrap())]
            .into_iter()
            .collect()
    }
}
