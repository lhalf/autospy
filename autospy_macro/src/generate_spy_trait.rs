use crate::associated_types::AssociatedSpyTypes;
use crate::generics::generics_idents;
use crate::inspect::cfg;
use crate::strip_attributes::{strip_attributes_from_signature, strip_autospy_attributes};
use crate::{attribute, edit, generate, inspect};
use proc_macro2::TokenStream;
use quote::{format_ident, quote};
use syn::{ItemTrait, Token, TraitItemConst, TraitItemFn, Type, parse_quote};

pub fn generate_spy_trait(
    item_trait: &ItemTrait,
    associated_spy_types: &AssociatedSpyTypes,
) -> TokenStream {
    let cfg = cfg();

    let trait_attributes = &item_trait.attrs;
    let trait_name = &item_trait.ident;

    let r#unsafe = &item_trait.unsafety;

    let generics = &item_trait.generics;
    let generics_where_clause = &generics.where_clause;
    let generics_idents = generics_idents(generics);

    let spy_name = format_ident!("{}Spy", trait_name);
    let associated_type_definitions = associated_type_definitions(associated_spy_types);
    let spy_associated_consts = spy_associated_consts(item_trait);
    let spy_function_definitions = trait_spy_function_definitions(item_trait);

    quote! {
        #cfg
        #(#trait_attributes)*
        #r#unsafe impl #generics #trait_name #generics_idents for #spy_name #generics_idents #generics_where_clause {
            #(#associated_type_definitions)*
            #(#spy_associated_consts)*
            #(#spy_function_definitions)*
        }
    }
}

fn associated_type_definitions(
    associated_spy_types: &AssociatedSpyTypes,
) -> impl Iterator<Item = TokenStream> {
    associated_spy_types
        .iter()
        .map(|(name, r#type)| quote! { type #name = #r#type; })
}

fn trait_spy_function_definitions(item_trait: &ItemTrait) -> impl Iterator<Item = TokenStream> {
    inspect::trait_functions(item_trait).map(function_as_spy_function)
}

fn function_as_spy_function(function: &TraitItemFn) -> TokenStream {
    let mut signature = function.sig.clone();

    if attribute::has_use_default_attribute(&function.attrs)
        && let Some(default_function) = &function.default
    {
        return quote! { #signature #default_function };
    }

    let function_name = &function.sig.ident;
    let spy_arguments = generate::tuple_or_single(
        inspect::spyable_arguments(function).map(argument_to_spy_expression),
    );

    edit::underscore_ignored_arguments_in_signature(&mut signature);
    strip_attributes_from_signature(&mut signature);

    quote! {
        #signature {
            self.#function_name.spy(#spy_arguments)
        }
    }
}

fn argument_to_spy_expression(argument: inspect::SpyableArgument) -> TokenStream {
    let argument_name = &argument.name;

    if let Some(with_expression) = argument.with_expression {
        return quote! { #with_expression ( #argument_name ) };
    }

    if argument.into_type.is_some() {
        return quote! { #argument_name.into() };
    }

    if let Type::ImplTrait(_) = argument.dereferenced_type {
        return quote! { Box::new(#argument_name) };
    }

    match argument.dereference_count {
        0 => quote! { #argument_name },
        1 => quote! { #argument_name.to_owned() },
        _ => {
            let dereferences = dereference_tokens(&argument);
            quote! { (#dereferences #argument_name).to_owned() }
        }
    }
}

fn dereference_tokens(argument: &inspect::SpyableArgument) -> TokenStream {
    "*".repeat((argument.dereference_count - 1) as usize)
        .parse()
        .expect("always valid token stream")
}

fn spy_associated_consts(item_trait: &ItemTrait) -> impl Iterator<Item = TokenStream> {
    inspect::associated_consts(item_trait).map(associated_const_as_spy_associated_const)
}

fn associated_const_as_spy_associated_const(associated_const: &TraitItemConst) -> TokenStream {
    let mut associated_const = associated_const.clone();

    associated_const.default = Some((
        <Token![=]>::default(),
        match attribute::associated_const(&associated_const.attrs) {
            None => {
                let associated_const_type = &associated_const.ty;
                parse_quote! { <#associated_const_type as const_default::ConstDefault>::DEFAULT }
            }
            Some(expr) => expr,
        },
    ));

    strip_autospy_attributes(&mut associated_const.attrs);

    quote! { #associated_const }
}

#[cfg(test)]
mod tests {
    use crate::associated_types::AssociatedSpyTypes;

    use super::generate_spy_trait;
    use quote::{ToTokens, quote};
    use syn::{ItemTrait, parse_quote};

    #[test]
    fn empty_generated_trait() {
        let input: ItemTrait = parse_quote! {
            trait Example {}
        };

        let expected = quote! {
            #[cfg(test)]
            impl Example for ExampleSpy {}
        };

        let actual = generate_spy_trait(&input, &AssociatedSpyTypes::new());

        assert_eq!(actual.to_token_stream().to_string(), expected.to_string());
    }

    #[test]
    fn empty_generated_trait_impl() {
        let input: ItemTrait = parse_quote! {
            trait Example {}
        };

        let expected = quote! {
            #[cfg(test)]
            impl Example for ExampleSpy {}
        };

        let actual = generate_spy_trait(&input, &AssociatedSpyTypes::new());

        assert_eq!(actual.to_token_stream().to_string(), expected.to_string());
    }

    #[test]
    fn trait_attributes_are_retained() {
        let input: ItemTrait = parse_quote! {
            #[some_attribute]
            trait Example {}
        };

        let expected = quote! {
            #[cfg(test)]
            #[some_attribute]
            impl Example for ExampleSpy {}
        };

        let actual = generate_spy_trait(&input, &AssociatedSpyTypes::new());

        assert_eq!(actual.to_token_stream().to_string(), expected.to_string());
    }

    #[test]
    fn async_trait_functions() {
        let input: ItemTrait = parse_quote! {
            #[async_trait]
            trait Example {
                async fn function(&self);
            }
        };

        let expected = quote! {
            #[cfg(test)]
            #[async_trait]
            impl Example for ExampleSpy {
                async fn function(&self) {
                    self.function.spy(())
                }
            }
        };

        let actual = generate_spy_trait(&input, &AssociatedSpyTypes::new());

        assert_eq!(actual.to_token_stream().to_string(), expected.to_string());
    }

    #[test]
    fn ignored_arguments_are_underscored_and_not_captured_in_trait_impl() {
        let input: ItemTrait = parse_quote! {
            trait Example {
                fn function(&self, #[autospy(ignore)] ignored: &str, captured: &str);
            }
        };

        let expected = quote! {
            #[cfg(test)]
            impl Example for ExampleSpy {
                fn function(&self, _: &str, captured: &str) {
                    self.function.spy(captured.to_owned())
                }
            }
        };

        let actual = generate_spy_trait(&input, &AssociatedSpyTypes::new());

        assert_eq!(actual.to_token_stream().to_string(), expected.to_string());
    }

    #[test]
    fn functions_with_static_impl_arguments_are_boxed() {
        let input: ItemTrait = parse_quote! {
            trait Example {
                fn function(&self, argument: impl ToString + 'static);
            }
        };

        let expected = quote! {
            #[cfg(test)]
            impl Example for ExampleSpy {
                fn function(&self, argument: impl ToString + 'static) {
                    self.function.spy(Box::new(argument))
                }
            }
        };

        let actual = generate_spy_trait(&input, &AssociatedSpyTypes::new());

        assert_eq!(actual.to_token_stream().to_string(), expected.to_string());
    }

    #[test]
    fn arguments_with_into_attribute_are_captured() {
        let input: ItemTrait = parse_quote! {
            trait Example {
                fn function(&self, #[autospy(into="IpAddr")] ip: [u8; 4]);
            }
        };

        let expected = quote! {
            #[cfg(test)]
            impl Example for ExampleSpy {
                fn function(&self, ip: [u8; 4]) {
                    self.function.spy(ip.into())
                }
            }
        };

        let actual = generate_spy_trait(&input, &AssociatedSpyTypes::new());

        assert_eq!(actual.to_token_stream().to_string(), expected.to_string());
    }

    #[test]
    fn associated_consts_can_be_substituted_by_attribute() {
        let input: ItemTrait = parse_quote! {
            trait Example {
                #[autospy(100)]
                const VALUE: u64;
            }
        };

        let expected = quote! {
            #[cfg(test)]
            impl Example for ExampleSpy {
                const VALUE: u64 = 100;
            }
        };

        let actual = generate_spy_trait(&input, &AssociatedSpyTypes::new());

        assert_eq!(actual.to_token_stream().to_string(), expected.to_string());
    }

    #[test]
    fn associated_consts_uses_const_default_if_no_attribute_specified() {
        let input: ItemTrait = parse_quote! {
            trait Example {
                const VALUE: u8;
            }
        };

        let expected = quote! {
            #[cfg(test)]
            impl Example for ExampleSpy {
                const VALUE: u8 = <u8 as const_default::ConstDefault>::DEFAULT;
            }
        };

        let actual = generate_spy_trait(&input, &AssociatedSpyTypes::new());

        assert_eq!(actual.to_token_stream().to_string(), expected.to_string());
    }

    #[test]
    fn associated_consts_with_multiple_attributes_retain_non_autospy_attributes() {
        let input: ItemTrait = parse_quote! {
            trait Example {
                #[autospy("hello")]
                #[some_attribute]
                const VALUE: &'static str;
            }
        };

        let expected = quote! {
            #[cfg(test)]
            impl Example for ExampleSpy {
                #[some_attribute]
                const VALUE: &'static str = "hello";
            }
        };

        let actual = generate_spy_trait(&input, &AssociatedSpyTypes::new());

        assert_eq!(actual.to_token_stream().to_string(), expected.to_string());
    }

    #[test]
    fn multiple_associated_consts_can_be_substituted_by_attribute() {
        let input: ItemTrait = parse_quote! {
            trait Example {
                #[autospy(100)]
                const VALUE1: u64;
                #[autospy(false)]
                const VALUE2: bool;
            }
        };

        let expected = quote! {
            #[cfg(test)]
            impl Example for ExampleSpy {
                const VALUE1: u64 = 100;
                const VALUE2: bool = false;
            }
        };

        let actual = generate_spy_trait(&input, &AssociatedSpyTypes::new());

        assert_eq!(actual.to_token_stream().to_string(), expected.to_string());
    }

    #[test]
    fn default_trait_functions_marked_with_use_default_use_default_trait_function() {
        let input: ItemTrait = parse_quote! {
            trait Example {
                #[autospy(use_default)]
                fn one(&self) -> u8 {
                    1
                }
            }
        };

        let expected = quote! {
            #[cfg(test)]
            impl Example for ExampleSpy {
               fn one(&self) -> u8 {
                    1
                }
            }
        };

        let actual = generate_spy_trait(&input, &AssociatedSpyTypes::new());

        assert_eq!(actual.to_token_stream().to_string(), expected.to_string());
    }

    #[test]
    fn trait_impl_is_generic_over_trait_generic() {
        let input: ItemTrait = parse_quote! {
            trait Example<T> {}
        };

        let expected = quote! {
            #[cfg(test)]
            impl<T> Example<T> for ExampleSpy<T> {}
        };

        let actual = generate_spy_trait(&input, &AssociatedSpyTypes::new());

        assert_eq!(actual.to_token_stream().to_string(), expected.to_string());
    }

    #[test]
    fn trait_impl_is_generic_over_multiple_trait_generics() {
        let input: ItemTrait = parse_quote! {
            trait Example<T, R> {}
        };

        let expected = quote! {
            #[cfg(test)]
            impl<T, R> Example<T, R> for ExampleSpy<T, R> {}
        };

        let actual = generate_spy_trait(&input, &AssociatedSpyTypes::new());

        assert_eq!(actual.to_token_stream().to_string(), expected.to_string());
    }

    #[test]
    fn trait_impl_is_generic_over_trait_generic_with_bounds() {
        let input: ItemTrait = parse_quote! {
            trait Example<T: Copy> {}
        };

        let expected = quote! {
            #[cfg(test)]
            impl<T: Copy> Example<T> for ExampleSpy<T> {}
        };

        let actual = generate_spy_trait(&input, &AssociatedSpyTypes::new());

        assert_eq!(actual.to_token_stream().to_string(), expected.to_string());
    }

    #[test]
    fn trait_impl_is_generic_over_multiple_trait_generics_with_bounds() {
        let input: ItemTrait = parse_quote! {
            trait Example<T: Copy, P: Clone> {}
        };

        let expected = quote! {
            #[cfg(test)]
            impl<T: Copy, P: Clone> Example<T,P> for ExampleSpy<T,P> {}
        };

        let actual = generate_spy_trait(&input, &AssociatedSpyTypes::new());

        assert_eq!(actual.to_token_stream().to_string(), expected.to_string());
    }

    #[test]
    fn trait_impl_is_generic_over_trait_generic_with_where_clause() {
        let input: ItemTrait = parse_quote! {
            trait Example<T> where T: Copy {}
        };

        let expected = quote! {
            #[cfg(test)]
            impl<T> Example<T> for ExampleSpy<T> where T: Copy {}
        };

        let actual = generate_spy_trait(&input, &AssociatedSpyTypes::new());

        assert_eq!(actual.to_token_stream().to_string(), expected.to_string());
    }

    #[test]
    fn trait_impl_is_unsafe_if_input_trait_is_unsafe() {
        let input: ItemTrait = parse_quote! {
            unsafe trait Example {}
        };

        let expected = quote! {
            #[cfg(test)]
            unsafe impl Example for ExampleSpy {}
        };

        let actual = generate_spy_trait(&input, &AssociatedSpyTypes::new());

        assert_eq!(actual.to_token_stream().to_string(), expected.to_string());
    }

    #[test]
    fn trait_functions_are_unsafe_if_input_trait_function_is_unsafe() {
        let input: ItemTrait = parse_quote! {
            trait Example {
                unsafe fn function(&self);
            }
        };

        let expected = quote! {
            #[cfg(test)]
            impl Example for ExampleSpy {
                unsafe fn function(&self) {
                    self.function.spy(())
                }
            }
        };

        let actual = generate_spy_trait(&input, &AssociatedSpyTypes::new());

        assert_eq!(actual.to_token_stream().to_string(), expected.to_string());
    }
}
