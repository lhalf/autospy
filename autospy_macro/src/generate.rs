use crate::associated_types::get_associated_types;
use crate::generate_spy_struct::generate_spy_struct;
use crate::generate_spy_trait::generate_spy_trait;
use crate::strip_attributes::strip_attributes;
use proc_macro2::TokenStream;
use quote::quote;
use syn::ItemTrait;

pub fn generate(item: TokenStream) -> TokenStream {
    let item_trait: ItemTrait = syn::parse2(item).expect("invalid trait definition");
    let associated_type = get_associated_types(&item_trait);
    let stripped_item_trait = strip_attributes(item_trait.clone());
    let spy_struct = generate_spy_struct(&item_trait, &associated_type);
    let spy_trait = generate_spy_trait(&item_trait, &associated_type);

    quote! {
        #stripped_item_trait
        #spy_struct
        #spy_trait
    }
}

pub fn tuple_or_single(mut items: impl Iterator<Item = TokenStream>) -> TokenStream {
    match (items.next(), items.next(), items) {
        (None, _, _) => quote! { () },
        (Some(first), None, _) => quote! { #first },
        (Some(first), Some(second), remainder) => quote! { ( #first , #second #(, #remainder)* ) },
    }
}

#[cfg(test)]
mod tests {
    use crate::generate::generate;
    use proc_macro2::TokenStream;
    use quote::quote;

    fn generate_pretty(tokens: TokenStream) -> String {
        let expanded = generate(tokens).to_string();
        prettyplease::unparse(&syn::parse_file(&expanded).unwrap())
    }

    #[test]
    fn arguments_marked_with_into_attribute_are_captured_as_that_type() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait MyTrait {
                fn function(&self, #[autospy(into="IpAddr")] ip: [u8; 4]);
            }
        }));
    }

    #[test]
    fn arguments_marked_with_with_attribute_are_captured_with_that_expression() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait MyTrait {
                fn function(
                    &self,
                    #[autospy(into="Result<String,Utf8Error>", with="String::from_utf8")] bytes: Vec<u8>,
                );
            }
        }));
    }

    #[test]
    fn arguments_marked_with_ignore_attribute_are_not_captured() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                fn function(&self, #[autospy(ignore)] ignored: &str, captured: &str);
            }
        }));
    }

    #[test]
    fn arguments_marked_with_multiple_attributes_retain_non_autospy_attributes() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                fn function(&self, #[some_attribute] #[autospy(ignore)] ignored: &str, captured: &str);
            }
        }));
    }

    #[test]
    fn multiple_ignored_arguments_are_not_captured() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                fn function(&self, #[autospy(ignore)] ignored1: &str, #[autospy(ignore)] ignored2: &str, captured1: &str, captured2: &str);
            }
        }));
    }

    #[test]
    fn method_with_no_arguments() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                fn function(&self);
            }
        }));
    }

    #[test]
    fn multiple_methods_with_no_arguments() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                fn function1(&self);
                fn function2(&self);
            }
        }));
    }

    #[test]
    fn trait_with_non_pub_crate_visibility() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                fn function(&self);
            }
        }));
    }

    #[test]
    fn trait_with_pub_visibility() {
        insta::assert_snapshot!(generate_pretty(quote! {
            pub trait TestTrait {
                fn function(&self);
            }
        }))
    }

    #[test]
    fn trait_with_pub_crate_visibility() {
        insta::assert_snapshot!(generate_pretty(quote! {
            pub(crate) trait TestTrait {
                fn function(&self);
            }
        }))
    }

    #[test]
    fn method_with_non_void_return_type() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                fn function(&self) -> bool;
            }
        }))
    }

    #[test]
    fn single_owned_argument() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                fn function(&self, argument: String);
            }
        }))
    }

    #[test]
    fn borrowed_argument_is_converted_to_owned_type() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                fn function(&self, argument: &str);
            }
        }))
    }

    #[test]
    fn multiple_nested_references_in_argument_converted_to_owned_type() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                fn function(&self, argument: &&&str);
            }
        }))
    }

    #[test]
    fn multiple_owned_arguments() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                fn function(&self, argument1: String, argument2: String);
            }
        }))
    }

    #[test]
    fn multiple_borrowed_arguments_converted_to_owned() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                fn function(&self, argument1: &str, argument2: &str);
            }
        }))
    }

    #[test]
    fn multiple_arguments_that_are_nested_references_converted_to_owned() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                fn function(&self, argument1: &&&&str, argument2: &&&str);
            }
        }))
    }

    #[test]
    fn single_static_impl_argument_converted_to_boxed_dyn() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                fn function(&self, argument: impl ToString + 'static);
            }
        }))
    }

    #[test]
    fn multiple_impl_bounds_static_argument_converted_to_boxed_dyn() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                fn function(&self, argument: impl ToString + Debug + 'static);
            }
        }))
    }

    #[test]
    fn functions_marked_with_return_attribute_have_their_return_types_changed() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                #[autospy(returns = "String")]
                fn function(&self) -> impl ToString;
            }
        }))
    }

    #[test]
    fn functions_marked_with_multiple_attributes_retain_non_autospy_attributes() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                #[some_attribute]
                #[autospy(returns = "String")]
                fn function(&self) -> impl ToString;
            }
        }))
    }

    #[test]
    fn traits_with_single_associated_type_attribute_return_expected_type() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                #[autospy(String)] type Item;
                fn function(&self) -> Self::Item;
            }
        }))
    }

    #[test]
    fn traits_with_single_associated_type_attribute_capture_expected_type() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                #[autospy(String)] type Item;
                fn function(&self, argument: Self::Item);
            }
        }))
    }

    #[test]
    fn traits_with_multiple_associated_types() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                #[autospy(String)]
                type Argument;
                #[autospy(String)]
                type Return;
                fn function(&self, argument: Self::Argument) -> Self::Return;
            }
        }))
    }

    #[test]
    fn async_trait() {
        insta::assert_snapshot!(generate_pretty(quote! {
            #[async_trait]
            trait TestTrait {
                async fn function(&self);
            }
        }))
    }

    #[test]
    fn async_trait_with_other_trait_requirements() {
        insta::assert_snapshot!(generate_pretty(quote! {
            #[async_trait]
            trait TestTrait: Send + Sync + 'static {
                async fn function(&self);
            }
        }))
    }
    #[test]
    fn traits_with_associated_consts() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                #[autospy("example")]
                const VALUE: &'static str;
                fn function(&self);
            }
        }))
    }
    #[test]
    fn traits_with_default_definitions() {
        insta::assert_snapshot!(generate_pretty(quote! {
            trait TestTrait {
                #[autospy(use_default)]
                fn function(&self) -> u8 {
                    1
                }
            }
        }))
    }
}
