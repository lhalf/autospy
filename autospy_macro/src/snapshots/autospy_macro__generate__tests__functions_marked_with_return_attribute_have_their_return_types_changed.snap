---
source: autospy_macro/src/generate.rs
assertion_line: 220
expression: "generate_pretty(quote!\n{\n    trait TestTrait\n    { #[autospy(returns = \"String\")] fn function(&self) -> impl ToString; }\n})"
---
trait TestTrait {
    fn function(&self) -> impl ToString;
}
#[cfg(any(test, not(feature = "test")))]
#[derive(Default, Clone)]
struct TestTraitSpy {
    pub function: autospy::SpyFunction<(), String>,
}
#[cfg(any(test, not(feature = "test")))]
impl TestTrait for TestTraitSpy {
    fn function(&self) -> impl ToString {
        self.function.spy(())
    }
}
